File: src/java1/algorithms/array/threeSum/ThreeSum.java
package threeSum;

import java.util.*;

public class ThreeSum {
    // Two pointer approach:- TC: O(n logn) + O(n*2) => O(n * 2), SC: O(1)
    private static List<List<Integer>> threeSum(int[] nums, int target) {
        List<List<Integer>> triplets = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int sum = nums[left] + nums[right] + nums[i];
                if (sum < target) {
                    left++;
                } else if (sum > target) {
                    right--;
                } else {

                    triplets.add(List.of(nums[i], nums[left], nums[right]));
                    while (left < right && nums[left] == nums[left + 1])
                        left++;
                    while (left < right && nums[right] == nums[right - 1])
                        right--;
                    left++;
                    right--;
                }
            }
        }
        return triplets;
    }

    public static void main(String[] args) {
        System.out.println(threeSum(new int[] { 3, 3, -1, -2, 0, -2, 0, -1, 0, -1 }, 0)); // [[-2,-1,3],[-2,0,2],[-1,0,1]]
        System.out.println(threeSum(new int[] { -1, 0, 1, 2, -1, -4 }, 0)); // [[-1,-1,2],[-1,0,1]]
        System.out.println(threeSum(new int[] { 0, 0, 0 }, 0)); // [[0,0,0]]
        System.out.println(threeSum(new int[] { 0, 1, 1 }, 0)); // []
        System.out.println(threeSum(new int[] { -2, 0, 0, 2, 2 }, 0)); // [[-2,0,2]]
        System.out.println(threeSum(new int[] { 1, 2, 3 }, 6)); // [[1,2,3]]
    }
}

File: test/arrays.py/three_sum.py
def three_sum(nums, target):
    nums.sort()

    triplets = []

    for i in range(len(nums)-2):
        if i > 0 and nums[i] == nums[i-1]:
            continue

        left, right = i + 1, len(nums) -1

        while left < right:
            sum = nums[i] + nums[left] + nums[right]

            if sum < target:
                left += 1
            elif sum > target:
                right -= 1
            else:
                triplets.append([nums[i], nums[left], nums[right]])

                left += 1
                while left < right and nums[left] == nums[left-1]:
                    left += 1

                right -=1
                while left < right and nums[right] == nums[right+1]:
                    right -= 1
    return triplets

# Test cases
if __name__ == "__main__":
    test_cases = [
        {"nums": [3, 3, -1, -2, 0, -2, 0, -1, 0, -1], "target": 0},
        {"nums": [1, 3, 5, 1], "target": 0},
        {"nums": [-4, -1, -1, 0, 1, 2], "target": 0},
        {"nums": [0, 0, 0], "target": 0},  # Edge case: all zeroes
    ]

    for i, test in enumerate(test_cases):
        print(f"Test Case {i + 1}: {three_sum(test['nums'], test['target'])}")
            